### PR TITLE
Card Editor UI Improvements

### PR DESCRIPTION
**Pull request by zsarnett**
## Proposed change

- Dialog height and width changes
	- Height and Width are now fixed. Card Editor never changes height when editing
- Move `Show Code Editor` to overflow menu for Card Editor and SubElement Editor
	- This is a move to simplify the UI of Lovelace Editors
- Remove Overflow on the Error card
- Add Delete to the overflow menu for SubElements
- ~~Change the UX for reordering Entities in the Entities Card Editor~~
	- ~~Now requires a .5s hold in order to move the row~~
- Change the appearance on Entity Rows in the Entities Card Editor
	- You can no longer edit the row entity without entering the SubElement editor for that row
- Adds Advanced Mode to the editor
	- Only shows the simple options for the main editor and adds advanced accordion for the rest
- Convert `<ha-settings-row>` to `<mwc-list-item>`
- Change `<ha-help-tooltip>` icon to the outlined version
- Remove padding from `<ha-expansion-panel>`
- Hide overflow on `<hui-error-card>`
- Remove the optional label from the theme input
- Add a new config element template that has a slot for advanced options
- Convert button and entities cards to new config element template
- Changed Header and Footer editor UI in the entities card to be more like Google Chrome Settings

image
image
image
image
7f6a4602f2559bc31865443a72f8b1f5

## Breaking Change

- Only show `Show Code Editor` option to Users in Advanced mode

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

## TODO

- [x] Fix the Add Row button so that it properly adds a new row without error
- [x] Find out why I have to click the Show Code Editor twice when it is first used
- [x] Clean up PR
- [x] Update all Config Elements with Advanced toggle

### CODE REVIEW

**Comment by thomasloven** [comment]
`Show code editor` never works on the first try for me...

**Comment by thomasloven** [comment]
It was more clear how to move entities before, and the new version.
And the ripple gets stuck.
Skärminspelning 2020-11-16 kl  15 00 56 mov

**Comment by zsarnett** [comment]
> `Show code editor` never works on the first try for me...

Yea that's in my todos... Not sure why

> It was more clear how to move entities before. The handles made it very discoverable.

I agree it was more discoverable. I make add an icon back because I agree it isn't the best experience

**Comment by bramkragten** [comment]
>Change the UX for reordering Entities in the Entities Card Editor
Now requires a .5s hold in order to move the row

How does a user know that?

**Comment by zsarnett** [comment]
67f3b5300101bff0c4f698572a56e28f

**Comment by zsarnett** [comment]
d5ea490e43c7346be7aa1f4146551055

**Comment by zsarnett** [comment]
7f6a4602f2559bc31865443a72f8b1f5

**Comment by spacegaier** [comment]
> 
> 
> d5ea490e43c7346be7aa1f4146551055

I am not seeing the benefit of the checkbox in the three dots menu. Is there any harm of always having the "Advanced options" visible, but collapsed initially?

**Comment by zsarnett** [comment]
Yes there is harm for non advanced users. The first screen grab is for advanced users. They will always see it. The checkbox will only exist for non advanced users

**Comment by zsarnett** [comment]
13482f01767f6e581b9ff4760aae3436

**Comment by thomasloven** [comment]
Why is it a setting in the editor and not connected to the profile setting?

**Comment by zsarnett** [comment]
It's both

File: src/panels/lovelace/editor/hui-config-element-template.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+export class HuiConfigElementTemplate extends LitElement {
+  @property({ type: Boolean }) public isAdvanced? = false;
+
+  protected render(): TemplateResult {
+    return html`
+      <slot></slot>
+      ${this.isAdvanced
+        ? html`
+            <ha-expansion-panel>
+              <span class="advanced-title" slot="title">Advanced options</span>
```
Comment (zsarnett) [inline src/panels/lovelace/editor/hui-config-element-template.ts:22]: Lokalize

File: src/panels/lovelace/editor/hui-element-editor.ts [hunk 2]
```diff
@@ -148,6 +154,10 @@ export abstract class HuiElementEditor<T> extends LitElement {
     this.GUImode = !this.GUImode;
   }
 
+  public remove() {
```
Comment (zsarnett) [inline src/panels/lovelace/editor/hui-element-editor.ts:157]:
```suggestion
  public remove(): void {
```

File: src/panels/lovelace/editor/hui-entities-card-row-editor.ts [hunk 3]
```diff
@@ -63,78 +64,69 @@ export class HuiEntitiesCardRowEditor extends LitElement {
...
-        )})`}
-      </h3>
+      <div class="title">
+        <span>
+          ${this.label ||
+          `${this.hass!.localize(
+            "ui.panel.lovelace.editor.card.generic.entities"
+          )}:`}
+        </span>
+        <mwc-button label="Add Row" @click=${this._addEntity}></mwc-button>
```
Comment (zsarnett) [inline src/panels/lovelace/editor/hui-entities-card-row-editor.ts:74]: Lokalize

File: src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts [hunk 4]
```diff
@@ -177,11 +210,75 @@ export class HuiDialogEditCard extends LitElement
...
-                      <ha-svg-icon .path=${mdiHelpCircle}></ha-svg-icon>
+                      <ha-svg-icon .path=${mdiHelpCircleOutline}></ha-svg-icon>
                     </mwc-icon-button>
                   </a>
                 `
               : ""}
+            ${this._cardConfig !== undefined
+              ? html`
+                  <ha-button-menu
+                    fixed
```
Comment (zsarnett) [inline src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts:221]: This is to account for a bug with MWC Menu in Dialogs

**Comment by MattiBaat** [comment]
Great job, nice improvements!

> Yes, there is harm for non-advanced users. The first screengrab is for advanced users. They will always see the advanced options. The checkbox will only exist for non-advanced users. Which can be managed on an admin's profile page

From this card perspective I understand that you want to show just the basics and hide more advanced settings. Somehow it doesn't feel right to overrule the advanced user option. In my opinion this will make the advanced user functionality less clear and decrease it's value. I would suggest to change the 'Advanced options' to 'More options' and be visible for every user. Or hide these advanced options for non advanced users.

**Comment by bramkragten** [comment]
@zsarnett Where are you!! We need you!! 😄

**Comment by zsarnett** [comment]
I know I know. Work is busy. No time for many hobbies right now. Different from the last pause. This time there is no other hobby just work... I am hoping the after hours work dies down soon so I can get back to dev and videos

**Comment by thomasloven** [comment]
Oh I know that feeling...
We're here when you come back.

**Pull request closed without merging**